REGRESSION(264795@main) media/video-object-fit.html test is failing

jernoble opened the pull request:
#### 9f0beb45c5d7c4830cd579eb68cee6564172c01c
<pre>
REGRESSION(264795@main) media/video-object-fit.html test is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=257604">https://bugs.webkit.org/show_bug.cgi?id=257604</a>
rdar://110117044

Reviewed by Eric Carlson.

In 264795@main, we now correctly lay out the contents of the media element when that element
recreates its media layer while in fullscreen or pip. This had the paradoxical effect of exposing
serious problems in our object-fit implementation for video elements.

The WebAVPlayerLayer class in the UI process is responsible for responding to layer-based layout
changes. These can happen both during normal page layout, and also when the video contents move
outside of the page and into fullscreen or pip. To correctly layout the video contents, it must
know what the effective videoGravity of the underlying video layer has been set to. However, this
information never leaves the WebContent process. This worked previously only accidentally.

Additionally, the RenderLayer is responsible for telling the MediaPlayer whether it should resize
respecting the underlying aspect ratio of the media, or whether it should ignore aspect ratio.
This code runs after layout, but changes to style that don&apos;t affect layout (like object-fit) are
never propogated into the media element.

And one final problem, WebAVPlayerLayer used pointer equality to check whether the passed in
string matches the values exposed by AVFoundation; it should use -isEqualToString:.

1. Notice that the object-fit style has changed in RenderVideo and push those changes into the
   MediaPlayer via updatePlayer().

2. Add a new videoGravity()/setVideoGravity() accessor to GraphicsLayer and GraphicsLayerCA, and
   to PlatformCALayer and PlatformCALayerCocoa.

3. Propogate these changes across the UIProcess boundary as part of the LayerProperties.

4. Fix WebAVPlayerLayer&apos;s -setVideoGravity: method.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer setVideoGravity:]):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::videoGravity const):
(WebCore::GraphicsLayer::setVideoGravity):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setVideoGravity):
(WebCore::GraphicsLayerCA::setContentsToVideoElement):
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateVideoGravity):
(WebCore::GraphicsLayerCA::layerChangeAsString):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::videoGravity const):
(WebCore::PlatformCALayerCocoa::setVideoGravity):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.h: Added.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.mm: Added.
(WebCore::convertAVLayerToMediaPlayerVideoGravity):
(WebCore::convertMediaPlayerToAVLayerVideoGravity):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::createPrimaryGraphicsLayer):
(WebCore::RenderLayerBacking::updateVideoGravity):
(WebCore::RenderLayerBacking::updateConfigurationAfterStyleChange):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderMedia.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::styleDidChange):
* Source/WebCore/rendering/RenderVideo.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerProperties::LayerProperties):
(WebKit::RemoteLayerTreeTransaction::LayerProperties::encode const):
(WebKit::RemoteLayerTreeTransaction::LayerProperties::decode):
(WebKit::dumpChangedLayers):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::updateClonedLayerProperties const):
(WebKit::PlatformCALayerRemote::videoGravity const):
(WebKit::PlatformCALayerRemote::setVideoGravity):

Canonical link: <a href="https://commits.webkit.org/264812@main">https://commits.webkit.org/264812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86fac9a6e848672303f2f220b222217647c56340

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8668 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8958 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9173 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10946 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8929 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11548 "1 flakes 98 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9830 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10481 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7130 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7925 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15453 "1 flakes 124 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8073 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11429 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6982 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7824 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2112 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8300 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -421,6 +422,24 @@ void GraphicsLayer::setMaskLayer(RefPtr<GraphicsLayer>&& layer)
+#if USE(CA)
+    return m_videoGravity;
+#else
+    return MediaPlayerVideoGravity::ResizeAspect;
```
`MediaPlayerVideoGravity::ResizeAspect` is the hard coded default gravity in a bunch of methods now so it might be worth defining a named constant.

jernoble replied:
One problem is that `ResizeAspect != 0`. If we are going to make a default value, we should also have it be equivalent to 0.

eric-carlson commented on Source/WebCore/rendering/RenderVideo.h:
```diff
@@ -76,6 +76,7 @@ class RenderVideo final : public RenderMedia {
     void paintReplaced(PaintInfo&, const LayoutPoint&) final;
 
     void layout() final;
+    void styleDidChange(StyleDifference, const RenderStyle* oldStyle) final;
```
Nit: `oldStyle` isn't necessary

eric-carlson commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -297,6 +299,11 @@ static void applyCommonPropertiesToLayer(CALayer *layer, const RemoteLayerTreeTr
 #endif
+
+    if (properties.changedProperties & LayerChange::VideoGravityChanged) {
+        if ([layer respondsToSelector:@selector(setVideoGravity:)])
+            [(WebAVPlayerLayer*)layer setVideoGravity:convertMediaPlayerToAVLayerVideoGravity(properties.videoGravity)];
```
Is it worth asserting `[layer isKindOfClass:[WebAVPlayerLayer  class]]`?

jernoble replied:
Maybe? A lot of classes respond to -setVideoGravity:. I almost wish there was an explicit protocol we could check for.

jernoble committed:
REGRESSION(264795@main) media/video-object-fit.html test is failing
https://bugs.webkit.org/show_bug.cgi?id=257604
rdar://110117044

Reviewed by Eric Carlson.

In 264795@main, we now correctly lay out the contents of the media element when that element
recreates its media layer while in fullscreen or pip. This had the paradoxical effect of exposing
serious problems in our object-fit implementation for video elements.

The WebAVPlayerLayer class in the UI process is responsible for responding to layer-based layout
changes. These can happen both during normal page layout, and also when the video contents move
outside of the page and into fullscreen or pip. To correctly layout the video contents, it must
know what the effective videoGravity of the underlying video layer has been set to. However, this
information never leaves the WebContent process. This worked previously only accidentally.

Additionally, the RenderLayer is responsible for telling the MediaPlayer whether it should resize
respecting the underlying aspect ratio of the media, or whether it should ignore aspect ratio.
This code runs after layout, but changes to style that don't affect layout (like object-fit) are
never propogated into the media element.

And one final problem, WebAVPlayerLayer used pointer equality to check whether the passed in
string matches the values exposed by AVFoundation; it should use -isEqualToString:.

1. Notice that the object-fit style has changed in RenderVideo and push those changes into the
   MediaPlayer via updatePlayer().

2. Add a new videoGravity()/setVideoGravity() accessor to GraphicsLayer and GraphicsLayerCA, and
   to PlatformCALayer and PlatformCALayerCocoa.

3. Propogate these changes across the UIProcess boundary as part of the LayerProperties.

4. Fix WebAVPlayerLayer's -setVideoGravity: method.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
(-[WebAVPlayerLayer setVideoGravity:]):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::videoGravity const):
(WebCore::GraphicsLayer::setVideoGravity):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setVideoGravity):
(WebCore::GraphicsLayerCA::setContentsToVideoElement):
(WebCore::GraphicsLayerCA::commitLayerChangesBeforeSublayers):
(WebCore::GraphicsLayerCA::updateVideoGravity):
(WebCore::GraphicsLayerCA::layerChangeAsString):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::videoGravity const):
(WebCore::PlatformCALayerCocoa::setVideoGravity):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.h: Added.
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.mm: Added.
(WebCore::convertAVLayerToMediaPlayerVideoGravity):
(WebCore::convertMediaPlayerToAVLayerVideoGravity):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::createPrimaryGraphicsLayer):
(WebCore::RenderLayerBacking::updateVideoGravity):
(WebCore::RenderLayerBacking::updateConfigurationAfterStyleChange):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderMedia.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::styleDidChange):
* Source/WebCore/rendering/RenderVideo.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerProperties::LayerProperties):
(WebKit::RemoteLayerTreeTransaction::LayerProperties::encode const):
(WebKit::RemoteLayerTreeTransaction::LayerProperties::decode):
(WebKit::dumpChangedLayers):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::updateClonedLayerProperties const):
(WebKit::PlatformCALayerRemote::videoGravity const):
(WebKit::PlatformCALayerRemote::setVideoGravity):

Canonical link: https://commits.webkit.org/264812@main
